### PR TITLE
port to python3

### LINKED ISSUE #221: Port codebase to python 3
Unfortunately we've reached a point where staying on python 2 is no longer a workable option. Our builds are fragile and our libraries buggy because upstream doesn't even bother to test python 2 compatability anymore. Most libraries are frozen at old versions.

We will need to be careful with our 2 -> 3 port as we work with binary data fairly often. Thankfully most of this binary data such as filenames are in a context where we can assume ascii only, so converting them to unicode is fine - we don't need to support non-utf-8 filepaths. Segment data will be truly binary though, and working with binary files around `subprocess.Popen` may be painful.

It would be nice if we could approach this issue piecemeal, but due to the `common` library that's not really doable. We would need to maintain a cross-compatible common, or else two copies, in order to work with some components in py2 and some in py3. In practice, it will be simpler to do a hard switch then test and deal with breakage - thankfully we don't need to worry about uptime outside of the run, so we have plenty of time.

### PR DESCRIPTION
**Pull request by ekimekim**
some basic testing has been done in downloader and restreamer,
but mostly this is just eyeballing and grepping for required changes.

Fixes #221

### CODE REVIEW

File: segment_coverage/segment_coverage/main.py [hunk 1]
```diff
@@ -232,7 +232,7 @@ def create_coverage_page(self, quality):
 		final_path = '{}_coverage.html'.format(path_prefix)
 		common.ensure_directory(temp_path)
 		with open(temp_path, 'w') as f:
-			f.write(html)
+			common.writeall(f.write, html)
```
Comment (chrusher) [inline segment_coverage/segment_coverage/main.py:235]: Do we need the write wrapper here when we are just writing text to a normal file?

File: backfiller/backfiller/main.py [hunk 2]
```diff
@@ -231,7 +231,7 @@ def __init__(self, base_dir, channels, qualities, static_nodes=[],
 	def stop(self):
 		"""Shut down all workers and stop backfilling."""
 		self.logger.info('Stopping')
-		for node in self.workers.keys():
+		for node in list(self.workers.keys()):
```
Comment (chrusher) [inline backfiller/backfiller/main.py:234]: Do we really need a list here?

**Comment by chrusher** [comment]
I have looked over your changes and I do not see any problems. I think the only way we are going to find any subtle bugs lurking in the code is by using the code so please merge when you are ready.

File: segment_coverage/segment_coverage/main.py (hunk 1 above)
Comment (ekimekim) [reply segment_coverage/segment_coverage/main.py:235]: No, but I felt it was safer to just do it everywhere.

File: backfiller/backfiller/main.py (hunk 2 above)
Comment (ekimekim) [reply backfiller/backfiller/main.py:234]: Yes, becasuse `self.workers` is concurrently modified while we're looping, so we need to make a copy of the list as it was at the start of the loop.

**Pull request merged**